src/coreclr/tools/dotnet-pgo/SPGO/SampleProfile.cs

```diff
@@ -16,69 +16,56 @@ namespace Microsoft.Diagnostics.Tools.Pgo
-            FlowGraph fg,
-            Dictionary<BasicBlock, long> samples,
-            Dictionary<BasicBlock, long> smoothedSamples,
-            Dictionary<(BasicBlock, BasicBlock), long> smoothedEdgeSamples)
+            FlowGraph fg)
         {
             MethodIL = methodIL;
             FlowGraph = fg;
-            Samples = samples;
-            SmoothedSamples = smoothedSamples;
-            SmoothedEdgeSamples = smoothedEdgeSamples;
         }
 
         public MethodIL MethodIL { get; }
         public FlowGraph FlowGraph { get; }
-        public Dictionary<BasicBlock, long> Samples { get; }
-        public Dictionary<BasicBlock, long> SmoothedSamples { get; }
-        public Dictionary<(BasicBlock, BasicBlock), long> SmoothedEdgeSamples { get; }
+        public IReadOnlyDictionary<BasicBlock, long> RawSamples => _rawSamples;
+        public IReadOnlyDictionary<BasicBlock, long> SmoothedSamples => _smoothedSamples;
+        public IReadOnlyDictionary<(BasicBlock, BasicBlock), long> SmoothedEdgeSamples => _smoothedEdgeSamples;
+        public long AttributedSamples { get; set; }
 
-        /// <summary>
-        /// Given pairs of runs (as relative IPs in this function), create a sample profile.
-        /// </summary>
-        public static SampleProfile CreateFromLbr(MethodIL il, FlowGraph fg, NativeToILMap map, IEnumerable<(uint fromRva, uint toRva, long count)> runs)
+        public bool TryAttributeSamples(int ilOffset, long count)
         {
-            Dictionary<BasicBlock, long> bbSamples = fg.BasicBlocks.ToDictionary(bb => bb, bb => 0L);
-            foreach ((uint from, uint to, long count) in runs)
-            {
-                foreach (BasicBlock bb in map.LookupRange(from, to).Select(fg.Lookup).Distinct())
-                {
-                    if (bb != null)
-                        bbSamples[bb] += count;
-                }
-            }
+            BasicBlock bb = FlowGraph.Lookup(ilOffset);
+            if (bb == null)
+                return false;
 
-            FlowSmoothing<BasicBlock> flowSmooth = new FlowSmoothing<BasicBlock>(bbSamples, fg.Lookup(0), bb => bb.Targets, (bb, isForward) => bb.Size * (isForward ? 1 : 50) + 2);
-            flowSmooth.Perform();
+            AttributeSamples(bb, count);
+            return true;
+        }
 
-            return new SampleProfile(il, fg, bbSamples, flowSmooth.NodeResults, flowSmooth.EdgeResults);
+        public void AttributeSamples(BasicBlock bb, long count)
+        {
+            Debug.Assert(FlowGraph.Lookup(bb.Start) == bb);
+            CollectionsMarshal.GetValueRefOrAddDefault(_rawSamples, bb, out _) += count;
+            AttributedSamples += count;
         }
 
-        /// <summary>
-        /// Given some IL offset samples into a method, construct a profile.
-        /// </summary>
-        public static SampleProfile Create(MethodIL il, FlowGraph fg, IEnumerable<int> ilOffsetSamples)
+        public void SmootheFlow()
```

Comment:
Nit ..?
```suggestion
        public void SmoothFlow()
```

Reply:
Today I learned that 'smooth' can also be used as a verb. :-)